v2/client.go

```diff
@@ -50,14 +53,35 @@ func NewClient(config *ClientConfiguration) (Client, error) {
 		EnableAlphaFeatures: config.EnableAlphaFeatures,
 		httpClient:          httpClient,
 	}
-	c.doRequestFunc = c.doRequest
 
 	if config.AuthConfig != nil {
-		if config.AuthConfig.BasicAuthConfig == nil {
-			return nil, errors.New("BasicAuthConfig is required if AuthConfig is provided")
+		if config.AuthConfig.BasicAuthConfig != nil {
+			c.doRequestFunc = func(req *http.Request) (*http.Response, error) {
+				req.SetBasicAuth(config.AuthConfig.BasicAuthConfig.Username, config.AuthConfig.BasicAuthConfig.Password)
+				return c.httpClient.Do(req)
+			}
+		} else if config.AuthConfig.OAuthConfig != nil {
+			oauthConfig := config.AuthConfig.OAuthConfig
```

Comment:
I would rather that this logic live in `prepareAndDo` rather than being injected in a factory method.